nifi-nar-bundles/nifi-standard-services/nifi-lookup-services-bundle/nifi-lookup-services/src/main/java/org/apache/nifi/lookup/CSVRecordLookupService.java

```diff
@@ -0,0 +1,232 @@
+                    final Map<String, Object> properties = new HashMap<>();
+                    record.toMap().forEach((k, v) -> {
+                        if (!lookupKeyColumn.equals(k)) {
+                            properties.put(k, v);
+                        }
+                    });
+                    cache.put(key, properties);
+                }
+
+                if (cache.isEmpty()) {
+                    logger.warn("Lookup table is empty after reading file: " + csvFile);
+                }
+            } finally {
+                lock.unlock();
+            }
+        }
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    @Override
+    protected void init(final ControllerServiceInitializationContext context) throws InitializationException {
+        final List<PropertyDescriptor> properties = new ArrayList<>();
+        properties.add(CSV_FILE);
+        properties.add(CSV_FORMAT);
+        properties.add(LOOKUP_KEY_COLUMN);
+        properties.add(IGNORE_DUPLICATES);
+        this.properties = Collections.unmodifiableList(properties);
+    }
+
+    @OnEnabled
+    public void onEnabled(final ConfigurationContext context) throws InitializationException, IOException, FileNotFoundException {
+        this.csvFile = context.getProperty(CSV_FILE).getValue();
+        this.csvFormat = CSVFormat.Predefined.valueOf(context.getProperty(CSV_FORMAT).getValue()).getFormat();
+        this.lookupKeyColumn = context.getProperty(LOOKUP_KEY_COLUMN).getValue();
+        this.ignoreDuplicates = context.getProperty(IGNORE_DUPLICATES).asBoolean();
+        this.watcher = new SynchronousFileWatcher(Paths.get(csvFile), new LastModifiedMonitor(), 30000L);
+        try {
+            loadCache();
+        } catch (final IllegalStateException e) {
+            throw new InitializationException(e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public Optional<Record> lookup(final Map<String, String> coordinates) throws LookupFailureException {
+        if (coordinates == null) {
+            return Optional.empty();
+        }
+
+        final String key = coordinates.get(KEY);
+        if (StringUtils.isBlank(key)) {
+            return Optional.empty();
+        }
+
+        try {
+            if (watcher != null && watcher.checkAndReset()) {
+                loadCache();
+            }
+        } catch (final IllegalStateException | IOException e) {
+            throw new LookupFailureException(e.getMessage(), e);
+        }
+
+        final Record lookupRecord;
+        Map<String, Object> recordMap = cache.get(key);
+        if (recordMap != null) {
+            List<RecordField> recordFields = new ArrayList<>(recordMap.size());
+            recordMap.forEach((k, v) -> recordFields.add(new RecordField(k, RecordFieldType.STRING.getDataType())));
+            final RecordSchema lookupRecordSchema = new SimpleRecordSchema(recordFields);
```

Comment:
I think this should be in loadCache and set the same way the main hash would be set under @markap14's suggestion.